_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+In this case, you can make these cycles atomic using the transaction support of the Kafka Producer or Kafka Streams APIs.
+
+The Kafka Producer API is low level, so you need to ensure the right configuration and to commit offsets within the scope of the transaction.
+A transaction should be aborted in case failure during processing, which means none of the messages from the transaction will be readable by transactional consumers.
+Instead, with the Kafka Streams API you can simply set `processing.guarantee=exactly_once_v2` to enable EOS on your existing topology.
+
+### Considerations
+
+Each producer must configure its own static and unique [`transactional.id`](https://kafka.apache.org/documentation/#producerconfigs_transactional.id) to enable the zombie fencing logic, which avoids message duplicates.
+The `transactional.id` is used to uniquely identify the same producer instance across process restarts, and it is associated with the producer session.
+When starting, the producer must call `initTransactions` one time, so the broker can tidy up the state associated with any previous incarnations of this producer, as identified by the `transactional.id`.
+Thereafter, any existing producers using the same `transactional.id` are considered zombies, and are fenced off as soon as they try to send data.
+
+> Before Kafka 2.6 the `transactional.id` had to be a static encoding of the input partitions in order to avoid ownership transfer between application instances during rebalances, that would invalidate the fencing logic.
+> This limitation was hugely inefficient, because an application instance couldn't reuse a single thread-safe producer instance, but had to create one for each input partition. 
+> It was fixed with by forcing the producer to send the consumer group metadata along with the offsets to commit ([KIP-447](https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics)).
+
+To guarantee message ordering, a given producer can have at most one open transaction (they are executed serially).
+Consumers with `isolation.level=read_committed` are not allowed to advance to offsets which include open transactions, while messages from aborted transactions are filtered out.
+
+The transaction coordinator is a module running inside every Kafka broker.
+Each coordinator owns a subset of the partitions in the internal `__transaction_state` topic (i.e. the partitions for which its broker is the leader).
+The coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.
+
+It is important to note that transactions are only supported inside a single Kafka cluster.
+If the transaction scope includes processors with externally observable side-effects, you would need to use an additional component such as `ChainedKafkaTransactionManager` from the SpringBoot project.
+This component is able to synchronize a database transaction with the Kafka transaction.
```

Comment:
Well I think that comes with some caveats, so maybe we should reword this claim to reflect what it actually does.

Reply:
I'll drop this reference.

The ChainedKafkaTransactionManager is deprecated. Spring still supports this, but I don't have much experience with that and it would require a dedicated post.

Instead, I'll mention the "transactional outbox pattern" (with a link), which addresses the same concern and it's implementation independent.